compiler/optimizer/OMRCFGSimplifier.cpp

```diff
@@ -333,6 +333,292 @@ bool OMR::CFGSimplifier::simplifyInstanceOfTestToCheckcast(bool needToDuplicateT
+            }
+         }
+      }
+
+   if (!patternMatched)
+      return false;
+
+   if (trace())
+      traceMsg(comp(), "   Matched with first %s in block_%d\n", isCompareLT ? "ificmplt" : "ificmpge",
+                  _block->getNumber());
+
+   if (nextCompareBlock->getFirstRealTreeTop() != nextCompareBlock->getLastRealTreeTop()) 
+      {
+      if (trace())
+         traceMsg(comp(), "   Abort simplifyBoundCheckWithThrowException : pattern not matched\n");
+      return false;
+      }
+
+   TR::Node *secondCompareNode = nextCompareBlock->getLastRealTreeTop()->getNode();
+
+   if (secondCompareNode->getOpCodeValue() != TR::ificmplt && secondCompareNode->getOpCodeValue() != TR::ificmpge) 
+      {
+      if (trace())
+         traceMsg(comp(), "   Abort simplifyBoundCheckWithThrowException : The second compare block does not contain ificmplt or ificmpge\n");
+      return false;
+      }
+
+   bool firstNodeRequiresDuplicate = false;
+   bool secondNodeRequiresDuplicate = false;
+   bool isLastBlockRetBlock = true;
+   patternMatched = false;
+   TR::Block *retBlock = NULL;
+   TR::Block *secondThrowBlock = NULL;
+   bool isSecondCompareLT = secondCompareNode->getOpCodeValue() == TR::ificmplt;
+
+   if (secondCompareNode->getNumChildren() > 1) 
+      {
+      TR::Node *firstChild = secondCompareNode->getFirstChild();
+      TR::Node *secondChild = secondCompareNode->getSecondChild();
+      
+      if (firstChild->getOpCodeValue() == TR::iload) 
+         {
+         if (secondChild->getOpCodeValue() == TR::iconst && secondChild->getConstValue() == 0) 
+            {
+            patternMatched = true;
+            secondThrowBlock = isSecondCompareLT ? secondCompareNode->getBranchDestination()->getNode()->getBlock() : nextCompareBlock->getNextBlock();
+            retBlock = isSecondCompareLT ? nextCompareBlock->getNextBlock() : secondCompareNode->getBranchDestination()->getNode()->getBlock();
+            isLastBlockRetBlock = !isSecondCompareLT;
+            firstChildBndChk = secondCompareNode->getFirstChild();
+            firstNodeRequiresDuplicate = true;
+            }
+         else if (secondChild->getOpCode().isIntegralLoadVar() || secondChild->getOpCode().isLoadIndirect())
+            {
+            patternMatched = true;
+            secondThrowBlock = isSecondCompareLT ? nextCompareBlock->getNextBlock() : secondCompareNode->getBranchDestination()->getNode()->getBlock();
+            retBlock = isSecondCompareLT ? secondCompareNode->getBranchDestination()->getNode()->getBlock() : nextCompareBlock->getNextBlock();
+            isLastBlockRetBlock = isSecondCompareLT;
+            secondChildBndChk = secondCompareNode->getSecondChild();
+            secondNodeRequiresDuplicate = true;
+            }
+         }
+      }
+
+   if (!patternMatched || !firstChildBndChk || !secondChildBndChk || firstChildBndChk == secondChildBndChk || throwBlock != secondThrowBlock) 
```

Comment:
I can't find the place where we make sure that the two if statements compare the **same value** to 0 and some limit. I afraid we will match incorrect patterns.

Reply:
Agreed, added check to verify if both ificmp operates on the same variable.